### ISSUE TITLE
cpuid-bool: ARM support

### ISSUE THREAD

**Issue by tarcieri** [RustCrypto/utils#378]
It would be nice to be able to do ARM feature detection.

A notable use case would be detecting the [ARMv8 Cryptography Extensions](https://github.com/RustCrypto/asm-hashes/issues/5#issuecomment-812890980).

**Comment by newpavlov** [comment]
Unfortunately, it's not possible to do properly without RFC 2725. AFAIK ARM does not have a dedicated CPUID-like (BTW RISC-V is similar in this regard, while it has a CPUID-like instruction, it requires privileged mode), on come CPUs it's possible to detect capabilities by reading system registers, but I am not sure about portability of such solution. Currently, the best solution is to rely on a platform-dependent feature detection as we do in `sha-1` and `sha2`. But I am not sure if we should place this code into `cpuid-bool`.

**Comment by tarcieri** [comment]
> on come CPUs it's possible to detect capabilities by reading system registers

Curious about this

**Comment by newpavlov** [comment]
See the top answer [here](https://stackoverflow.com/questions/64864035/any-cpuid-like-instruction-in-armv8).

**Comment by tarcieri** [comment]
These seem potentially relevant:

> `IID_AA64_ISAR{0,1}_EL1` have information about which instructions are implemented on the device: CRC. SHA, Atomic, random numbers, etc.

**Comment by newpavlov** [comment]
Can you check if it works in user space on your M1 Mac? Later I can check your snippet on a RPi4 board.

**Comment by tarcieri** [comment]
Sure, I can give it a try

**Comment by tarcieri** [comment]
It looks like if we were to use the libc crate (or potentially emit our own system call instruction), we could use `libc::getauxval(AT_HWCAP)` to obtain the data from these registers in userspace:

https://www.kernel.org/doc/html/latest/arm64/elf_hwcaps.html

Namely:

> - **HWCAP_AES**: Functionality implied by ID_AA64ISAR0_EL1.AES == 0b0001.
> - **HWCAP_SHA1**: Functionality implied by ID_AA64ISAR0_EL1.SHA1 == 0b0001.
> - **HWCAP_SHA2**: Functionality implied by ID_AA64ISAR0_EL1.SHA2 == 0b0001.

**Comment by newpavlov** [comment]
>It looks like if we were to use the libc crate (or potentially emit our own system call instruction), we could use libc::getauxval(AT_HWCAP) to obtain the data from these registers in userspace on Linux

Yes, it's what I meant by "platform-dependent feature detection" and it's [exactly that we currently do](https://github.com/RustCrypto/hashes/blob/master/sha1/src/compress/aarch64.rs#L4-L8) in `sha1` and `sha-2`.

**Comment by tarcieri** [comment]
Aha! Neat!

What do you think about moving that functionality into this crate, and having an abstraction that works across both Linux and macOS?

**Comment by newpavlov** [comment]
I think `cpuid-bool` is a bit misleading name for a crate with ARM target feature detection support, but I guess creating a separate crate or renaming it is not worth the trouble. Though I am not sure about hard-coding "ARM + macOS = Crypto extension support". Can you check if the `AT_HWCAP` works on macOS or if there is a similar solution?

**Comment by tarcieri** [comment]
Here's a guide:

https://developer.apple.com/documentation/apple-silicon/addressing-architectural-differences-in-your-macos-code

They suggest using `sysctl(3)` which appears to be available via the `libc` crate.

It looks like the feature detection flags reside in the `hw.optional` namespace:

```
hw.optional.floatingpoint: 1
hw.optional.watchpoint: 4
hw.optional.breakpoint: 6
hw.optional.neon: 1
hw.optional.neon_hpfp: 1
hw.optional.neon_fp16: 1
hw.optional.armv8_1_atomics: 1
hw.optional.armv8_crc32: 1
hw.optional.armv8_2_fhm: 1
hw.optional.armv8_2_sha512: 1
hw.optional.armv8_2_sha3: 1
hw.optional.amx_version: 2
hw.optional.ucnormal_mem: 1
hw.optional.arm64: 1
```

Somewhat curiously all the ones we actually care about aren't here... I'm not sure if that suggests they're non-optional features on ARM64 or what but that's all I can assume.

**Comment by tarcieri** [comment]
@newpavlov 

> I think cpuid-bool is a bit misleading name for a crate with ARM target feature detection support, but I guess creating a separate crate or renaming it is not worth the trouble.

FWIW I just registered the `cpufeatures` crate if you'd like a more generic name.

**Comment by newpavlov** [comment]
>This comment in particular suggests it is and will always be safe to assume aes, sha1, and sha2 are available on Apple aarch64 targets, citing clang/llvm contributions by Apple engineers as precedent.

In this case I think Rust should enable the relevant target features itself for these targets, similarly to how SSE2 is enabled by default for `x86_64-*` targets. It means that the proposed fallback to compile-time detection outside of Linux should do the thing for us.

**Comment by newpavlov** [comment]
>ideally I think we shouldn't need to gate CPU feature detection on OS

Yes and it's exactly the purpose of RFC 2725.

>and the OS specifics should be handled in this crate

No, ideally it should be handled either by `std` or by third-party platform-specific crates à la allocator crates.

**Comment by tarcieri** [comment]
Well sure, this crate is a stopgap. But it doesn't seem like there has been a whole lot of progress on RFC 2725?
